### ISSUE TITLE
resolve.conf issue with installing v 2.0.21

### ISSUE THREAD

**Issue by mehranrasulian** [cpriego/valet-linux#104]
After upgrading valet to v2.0.21, it will lock the `/etc/resolve.conf` file and I can't edit this file either by root or normal user.
 vim gives me this error message: `E45: 'readonly' option is set (add ! to override)`.

It's the second time I reinstalled ubuntu because there wasn't any way to fix this issue (at least with my knowledge and researches).

So please remove this version before this problem happen to more people :(

Thank you to taking the time on this tool :+1:

**Comment by cpriego** [comment]
The new valet linux version works by directly editing the `/etc/resolv.conf` file (which honestly few people edit directly since there are a ton of quirks). Since that file is usually owned and controlled by either, resolvconf, openresolv, network-manager, systemd-resolved, dhcpd, etc. (a LOT of options really) the only choice I had to make a distro/conf agnostic tool was to rewrite and lock that file (as it is suggested in many many guides like arch-wiki and ubuntu official documentation), and read all of the nameservers AND resolv.conf files that those other programs generate. 

The solution to edit that file is quite simple, just remove the write lock:
```
chattr -i /etc/resolv.conf
```

BTW all of the changes are quite visible in the source code.

**Comment by mehranrasulian** [comment]
> The solution to edit that file is quite simple, just remove the write lock:
chattr -i /etc/resolv.conf

Thank you to respond. I'll test this, at a virtual os, at least. but the last time I got the `operation not permitted` error when editing, remove or change the permissions of the file, as the root and normal user

**Comment by jfelipeara** [comment]
after updating valet i can't navigate in the browser, i got the error **dns probe finished bad config**. i had to uninstall it

**Comment by cpriego** [comment]
@jfelipeara I have some questions:
- What distro and version are you using?
- Did you, at any given point used the **dev-dnsmasq** branch?
- What is the output of `systemctl status dnsmasq`

Thanks in advance for the help

**Comment by jfelipeara** [comment]
I think i managed to fix the problem, i believe i was using de google dns. I had some nameservers on `/etc/resolv.conf`, after installing valet there only was 127.0.0.1 on that file, so i copy and paste the nameservers that i had and paste it on `/etc/resolvconf/resolv.conf.d/head` file and my internet was restored
- i'm using Ubuntu 16.04.1 LTS
- i dont think i used the dev-dnsmasq
- the output, after i did the above, was this:
`dnsmasq.service - dnsmasq - A lightweight DHCP and caching DNS server
   Loaded: loaded (/lib/systemd/system/dnsmasq.service; enabled; vendor preset: 
  Drop-In: /run/systemd/generator/dnsmasq.service.d
           └─50-dnsmasq-$named.conf, 50-insserv.conf-$named.conf
   Active: active (running) since lun 2017-09-25 14:22:48 COT; 5min ago
  Process: 1392 ExecStartPost=/etc/init.d/dnsmasq systemd-start-resolvconf (code
  Process: 1094 ExecStart=/etc/init.d/dnsmasq systemd-exec (code=exited, status=
  Process: 985 ExecStartPre=/usr/sbin/dnsmasq --test (code=exited, status=0/SUCC
 Main PID: 1384 (dnsmasq)
    Tasks: 1
   Memory: 1.1M
      CPU: 65ms
   CGroup: /system.slice/dnsmasq.service
           └─1384 /usr/sbin/dnsmasq -x /var/run/dnsmasq/dnsmasq.pid -u dnsmasq -

sep 25 14:22:48 jfelipeara-PC dnsmasq[1384]: using nameserver 200.13.249.101#53
sep 25 14:22:48 jfelipeara-PC dnsmasq[1384]: using nameserver 200.31.208.101#53
sep 25 14:22:48 jfelipeara-PC dnsmasq[1392]: /etc/resolvconf/update.d/libc: Warn
sep 25 14:22:48 jfelipeara-PC dnsmasq[1384]: reading /opt/valet-linux/dns-server
sep 25 14:22:48 jfelipeara-PC dnsmasq[1384]: using nameserver 200.13.249.101#53
sep 25 14:22:48 jfelipeara-PC dnsmasq[1384]: using nameserver 200.31.208.101#53
sep 25 14:22:48 jfelipeara-PC systemd[1]: Started dnsmasq - A lightweight DHCP a
sep 25 14:22:48 jfelipeara-PC dnsmasq[1384]: reading /opt/valet-linux/dns-server
sep 25 14:22:48 jfelipeara-PC dnsmasq[1384]: using nameserver 200.13.249.101#53
sep 25 14:22:48 jfelipeara-PC dnsmasq[1384]: using nameserver 200.31.208.101#53`

**Comment by cpriego** [comment]
The output is the expected one.

Indeed, the new valet version overrides the `/etc/resolv.conf` file as is the most trustworthy method of assigning dns servers.

1. Valet searches for ALL `resolv.conf` files in /run (so it gets a list of all dns servers obtained via NetworkManager/resolvconf/openresolv/dhclient/etc) and saves them in /opt/valet/resolv.conf
2. It makes a backup of `/etc/resolv.conf`
3. It overwrites the content of `/etc/resolv.conf` to have only search domains and the 127.0.0.1 address (corresponding to the dnsmasq binding address)
4. Configures dnsmasq to forward dns queries to the servers found in step 1.

So if you directly modified `/etc/resolv.conf` instead of adding custom dns servers through your dhcp client those servers will be lost.

**Comment by jfelipeara** [comment]
@cpriego thank you very much. one more thing, when i update valet everything worked perfect, i had internet acces, but when i reboot i hadn't internet access. what do you think the problem was?

**Comment by cpriego** [comment]
Perhaps there was no nameservers detected. You could confirm that by reading the contents of /opt/valet/resolv.conf

**Comment by dikiaap** [comment]
To be honest I don't really understand the dns stuff here
I'm using Ubuntu 16.04 LTS and recently upgraded valet to v2.0.21.
I did a fresh install again. `valet uninstall`,` valet install`.
All look good. No issues. But after reboot. I can not connect to the internet.

Then I added Google Public DNS in `/etc/resolv.conf`
```
$ cat /etc/resolv.conf 
nameserver 127.0.0.1
nameserver 8.8.4.4
nameserver 8.8.8.8
```
and other similar ways by adding
```
server=8.8.8.8
server=8.8.4.4
no-resolv
```
in `/etc/dnsmasq.conf`

It only works through tethering from my phone.
But it doesn't work via public wifi that has login page.
I did that `chattr -i resolv.conf`, it also doesn't work.

**Comment by cpriego** [comment]
@dikiaap I myself I'm using 16.04 so the testing there is more intense. It looks like something did not installed correctly.

Could you share the output of the following commands please?
```
ls /opt/valet

cat /opt/valet/resolv.conf

cat /etc/dnsmasq.conf | grep conf.d

ps aux | grep inotify

systemctl status dnsmasq
```

**Comment by dikiaap** [comment]
@cpriego 

```shell
# It's valet-linux here, not valet
$ ls /opt | grep valet
valet-linux
$ ls /opt/valet-linux 
dns-servers  get-dns-servers  watch.log
```
```shell
$ cat /etc/rc.local
#!/bin/sh -e
#
# rc.local
#
# This script is executed at the end of each multiuser runlevel.
# Make sure that the script will "exit 0" on success or any other
# value on error.
#
# In order to enable or disable this script just change the execution
# bits.
#
# By default this script does nothing.

/opt/valet-linux/get-dns-servers

exit 0
```
```shell
$ cat /opt/valet-linux/resolv.conf
2017/09/27 12:24:39 open /opt/valet-linux/resolv.conf: no such file or directory
```
```shell
$ cat /etc/dnsmasq.conf | grep conf.d
conf-dir=/etc/dnsmasq.d
#conf-dir=/etc/dnsmasq.d,.bak
#conf-dir=/etc/dnsmasq.d/,*.conf
```
```shell
$ ps aux | grep -v grep | grep inotifywait
root       950  0.0  0.0   2332   600 ?        S    12:05   0:00 inotifywait -q -m -e modify --format %w%f /run/dnsmasq
root       952  0.0  0.0   2332   592 ?        S    12:05   0:00 inotifywait -q -m -e modify --format %w%f /run/resolvconf
root       959  0.0  0.0   2332   604 ?        S    12:05   0:00 inotifywait -q -m -e modify --format %w%f /run/resolvconf/interface
```
```shell
$ systemctl status dnsmasq
● dnsmasq.service - dnsmasq - A lightweight DHCP and caching DNS server
   Loaded: loaded (/lib/systemd/system/dnsmasq.service; enabled; vendor preset: enabled)
  Drop-In: /run/systemd/generator/dnsmasq.service.d
           └─50-dnsmasq-$named.conf, 50-insserv.conf-$named.conf
   Active: active (running) since Wed 2017-09-27 12:05:13 WIB; 20min ago
  Process: 789 ExecStartPost=/etc/init.d/dnsmasq systemd-start-resolvconf (code=exited, status=0/SUCCESS)
  Process: 738 ExecStart=/etc/init.d/dnsmasq systemd-exec (code=exited, status=0/SUCCESS)
  Process: 691 ExecStartPre=/usr/sbin/dnsmasq --test (code=exited, status=0/SUCCESS)
 Main PID: 788 (dnsmasq)
   CGroup: /system.slice/dnsmasq.service
           └─788 /usr/sbin/dnsmasq -x /var/run/dnsmasq/dnsmasq.pid -u dnsmasq -7 /etc/dnsmasq.d,.dpkg-dist,.dpkg-old,.dpkg-new --local-service

Sep 27 12:05:08 system systemd[1]: Starting dnsmasq - A lightweight DHCP and caching DNS server...
Sep 27 12:05:10 system dnsmasq[691]: dnsmasq: syntax check OK.
Sep 27 12:05:13 system dnsmasq[788]: started, version 2.75 cache disabled
Sep 27 12:05:13 system dnsmasq[788]: compile time options: IPv6 GNU-getopt DBus i18n IDN DHCP DHCPv6 no-Lua TFTP conntrack ipset auth DNSSEC loop-detect inotify
Sep 27 12:05:13 system dnsmasq[788]: no servers found in /opt/valet-linux/dns-servers, will retry
Sep 27 12:05:13 system dnsmasq[788]: read /etc/hosts - 143 addresses
Sep 27 12:05:13 system dnsmasq[789]: /etc/resolvconf/update.d/libc: Warning: /etc/resolv.conf is not a symbolic link to /run/resolvconf/resolv.conf
Sep 27 12:05:13 system systemd[1]: Started dnsmasq - A lightweight DHCP and caching DNS server.
```

**Comment by santigarcor** [comment]
@cpriego I'm having the same issue, now i can't reinstall valet because if i reinstall it, i'll lose my internet access. Here's my output to the commands you posted having valet installed:

```
$ ls /opt/valet-linux

dns-servers  get-dns-servers  watch.log
```

```
$ cat /etc/rc.local

#!/bin/sh -e
#
# rc.local
#
# This script is executed at the end of each multiuser runlevel.
# Make sure that the script will "exit 0" on success or any other
# value on error.
#
# In order to enable or disable this script just change the execution
# bits.
#
# By default this script does nothing.

/opt/valet-linux/get-dns-servers

exit 0
```

cat /opt/valet-linux/resolv.conf
This files does not exist

```
$ cat /etc/dnsmasq.conf | grep conf.d

conf-dir=/etc/dnsmasq.d
#conf-dir=/etc/dnsmasq.d,.bak
#conf-dir=/etc/dnsmasq.d/,*.conf
```

```
$ ps aux | grep -v grep | grep inotifywait

root     10526  0.0  0.0   6532   704 pts/2    S    10:45   0:00 inotifywait -q -m -e modify --format %w%f /run/NetworkManager
root     10527  0.0  0.0   6532   688 pts/2    S    10:45   0:00 inotifywait -q -m -e modify --format %w%f /run/dnsmasq
root     10530  0.0  0.0   6532   852 pts/2    S    10:45   0:00 inotifywait -q -m -e modify --format %w%f /run/resolvconf
root     10536  0.0  0.0   6532   692 pts/2    S    10:45   0:00 inotifywait -q -m -e modify --format %w%f /run/resolvconf/interface
```

```
$ systemctl status dnsmasq

dnsmasq.service - dnsmasq - A lightweight DHCP and caching DNS server
   Loaded: loaded (/lib/systemd/system/dnsmasq.service; enabled; vendor preset: enabled)
  Drop-In: /run/systemd/generator/dnsmasq.service.d
           └─50-dnsmasq-$named.conf, 50-insserv.conf-$named.conf
   Active: failed (Result: exit-code) since jue 2017-09-28 10:45:14 COT; 3min 25s ago
  Process: 5231 ExecStop=/etc/init.d/dnsmasq systemd-stop-resolvconf (code=exited, status=0/SUCCESS)
  Process: 1200 ExecStartPost=/etc/init.d/dnsmasq systemd-start-resolvconf (code=exited, status=0/SUCCESS)
  Process: 10678 ExecStart=/etc/init.d/dnsmasq systemd-exec (code=exited, status=2)
  Process: 10675 ExecStartPre=/usr/sbin/dnsmasq --test (code=exited, status=0/SUCCESS)
 Main PID: 1199 (code=exited, status=0/SUCCESS)

sep 28 10:45:14 systemd[1]: Starting dnsmasq - A lightweight DHCP and caching DNS server...
sep 28 10:45:14 dnsmasq[10675]: dnsmasq: syntax check OK.
sep 28 10:45:14 dnsmasq[10678]: dnsmasq: failed to create listening socket for 127.0.0.1: Address already in use
sep 28 10:45:14 systemd[1]: dnsmasq.service: Control process exited, code=exited status=2
sep 28 10:45:14 systemd[1]: Failed to start dnsmasq - A lightweight DHCP and caching DNS server.
sep 28 10:45:14 systemd[1]: dnsmasq.service: Unit entered failed state.
sep 28 10:45:14 systemd[1]: dnsmasq.service: Failed with result 'exit-code'.
```

**Comment by rodrigopedra** [comment]
I downgraded valet-linux to 2.0.20 on Ubuntu 16.04.3 LTS so I could get the internet to work again

**Comment by cpriego** [comment]
This should be fixed in [v2.0.22](https://github.com/cpriego/valet-linux/releases/tag/v2.0.22)

**Comment by rodrigopedra** [comment]
I can confirm it is working on Ubuntu 16.04.3 LTS. But it only worked after a full restart.

Thanks for the good work!
